app/services/UIBuilder.scala

```diff
@@ -28,27 +28,35 @@ import scala.annotation.tailrec
 class UIBuilder {
   val logger: Logger = Logger(getClass)
 
-  def fromStanzas(url: String, stanzas: Seq[Stanza], formData: Option[FormData] = None)(implicit stanzaIdToUrlMap: Map[String, String]): Page =
-    Page(url,
-         stanzas.foldLeft(Seq[UIComponent]()) {(acc, stanza) =>
-           stanza match {
-             case i: Instruction => acc :+ fromInstruction(i)
-             case ig: InstructionGroup => acc :+ fromInstructionGroup(ig)
-             case c: Callout => acc ++ fromCallout(c, formData)
-             case rg: RowGroup => acc :+ fromRowGroup(rg)
-             case in: OcelotInput => Seq(fromInput(in, formData, acc))
-             case q: OcelotQuestion => Seq(fromQuestion(q, acc))
-             case _ => acc
-           }
-         })
-
-  private def fromRowGroup(rg: RowGroup)(implicit stanzaIdToUrlMap: Map[String, String]): UIComponent = {
-    val rowLength = rg.group.map(_.cells.length).max
-    SummaryList(rg.group.map{row =>
-      SummaryRow((row.cells ++ Seq.fill(rowLength - row.cells.size)(Phrase())).map{phrase =>
-        SummaryCell(TextBuilder.fromPhrase(phrase))
-      })
-    })
+  def buildPage(url: String, stanzas: Seq[VisualStanza], formData: Option[FormData] = None)
+               (implicit stanzaIdToUrlMap: Map[String, String]): Page =
+    Page(url, fromStanzas(stackStanzas(stanzas, Nil), Nil, formData))
+
+  @tailrec
+  private def fromStanzas(stanzas: Seq[VisualStanza], acc: Seq[UIComponent], formData: Option[FormData])
+                 (implicit stanzaIdToUrlMap: Map[String, String]): Seq[UIComponent] =
+    stanzas match {
+      case Nil => acc
+      case (sg: StackedGroup) :: xs => fromStanzas(xs, acc ++ fromStackedGroup(sg, formData), formData)
+      case (i: Instruction) :: xs => fromStanzas(xs, acc ++ Seq(fromInstruction(i)), formData)
+      case (ig: InstructionGroup) :: xs => fromStanzas(xs, acc ++ Seq(fromInstructionGroup(ig)), formData)
+      case (rg: RowGroup) :: xs if rg.isSummaryList =>
```

Comment:
Two questions:

1) what happens if the row group is not a summary list? Is it ignored or should that situation be caught in fromStackedGroup (which I haven't looked at yet!)
2) Is it necessary to invoke stackStanzas again when processing a summary list

Reply:
1) At the moment, its ignored as not corresponding to a GDS pattern
2) No, well spotted, will change